### PR TITLE
feat: new UX for the boards/library manager widget

### LINKED ISSUE #19: Make Boards/Library Manager uninstall functionality easier to find
### Describe the problem

I needed to uninstall a boards platform via Boards Manager and couldn't find the <kbd>**Remove**</kbd> button I'm used to from using Arduino IDE 1.x.

It took me quite some time to realize that the "**INSTALLED**" label on the entry in the list of platforms turns into an "**UNINSTALL**" button when you hover on it.

### To reproduce

1. Open the "**Boards Manager**" view.
1. Scroll down until you see the "**Arduino AVR Boards**" platform entry.
1. Look for an obvious mechanism to uninstall the platform. 

### Expected behavior

Make the uninstall functionality easy for the user to discover.

#### Original report

2.0.0-beta.2-snapshot.298cc11

#### Last verified with

76f9f635d8558f399423f64c0c709e3a53bdcd03

### Operating system

All

### Operating system version

Any

### Additional context

A minimal change that would improve the current situation would be to change the installed "**INSTALLED**" label to have the same border and drop shadow as the <kbd>INSTALL</kbd> button.

Alternatively, a <kbd>**Remove**</kbd> button could be added to mirror the Arduino IDE 1.x Boards Manager UI.

---

The Arduino IDE 1.x Library Manager does not have an uninstall capability, but the Arduino IDE 1.x Boards Manager does have an uninstall capability, so the latter should be used if looking for a reference design.

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://www.arduino.cc/en/software#nightly-builds)
- [X] My request contains all necessary details

### LINKED ISSUE #781: Library/Boards Manager offers already installed release when it is the only one
## Describe the bug

When multiple releases of a library or boards platform are available, Library and Boards Manager offer only the available versions other than the currently installed one.

:bug: When there is only one available release of a library or boards platform, the user is still offered an <kbd>INSTALL</kbd> button. If they click it, the IDE attempts to install the version that is already installed.

## To Reproduce

1. Open the Arduino IDE Library Manager.
1. Search for `Arduino Uno WiFi Dev Ed Library`
1. Install "**Arduino Uno WiFi Dev Ed Library by Arduino**".
   :bug: Notice that the library listing still has an <kbd>INSTALL</kbd> button for the already installed 0.0.3 release:
   image
1. Click the <kbd>INSTALL</kbd> button.
   :bug: Notice that the version shown in the notifications is identified as `undefined`
   ```
   Processing Arduino Uno WiFi Dev Ed Library:undefined: Already installed Arduino Uno WiFi Dev Ed Library@0.0.3
   ```
   ```
   Successfully installed library Arduino Uno WiFi Dev Ed Library:undefined
   ```
   image
1. Hover the mouse pointer over the "**INSTALLED**" label on the "**Arduino Uno WiFi Dev Ed Library by Arduino**" library listing.
   (It turns into an <kbd>UNINSTALL</kbd> button)
1. Click the <kbd>UNINSTALL</kbd> button.
   - :bug: Notice that the "**Arduino Uno WiFi Dev Ed Library by Arduino**" library listing still has an "**INSTALLED**" label, even though the library was really uninstalled.
   - :bug: Notice that the **Sketch > Include Library** menu still contains "**Arduino_Uno_WiFi_Dev_Ed_Library**".
1. Open the Arduino IDE Boards Manager.
1. Search for `littleBits`
1. Install "**littleBits Arduino AVR Modules by littleBits Electronics**".
   :bug: Notice that the platform listing still has an <kbd>INSTALL</kbd> button for the already installed 1.0.0 release:
   image

## Expected behavior

Remove or disable the <kbd>INSTALL</kbd> button when there is nothing available to install.

This is how it has always worked in Arduino IDE 1.x:

image

image

## Desktop

- OS: Windows 10
- Version: 2.0.0-rc3-snapshot.1d88263
Date: 2022-01-24T16:22:30.529Z
CLI Version: 0.20.2 [13783819]

## Additional context

In addition to libraries and boards platforms that only had one release, such as the ones used in the demo above, this condition will be present for all installed boards platforms without a package index. Because the "**Additional Boards Manager URLs**" preference that provides these package indexes is not imported from the Arduino IDE 1.x preferences, that will be the case for all 3rd party boards platforms present on the system of a user who has just updated to Arduino IDE 2.x, even in the session during which they have manually updated that preference (https://github.com/arduino/arduino-ide/issues/637). So it is not a rare condition.

---

I did not experience the same problem with the uninstall not being recognized after the double install in Boards Manager as I did in Library Manager.

---

I missed a previous report of the same bug at https://github.com/arduino/arduino-ide/issues/157. The discussion there might provide additional information if the problem is not clear from what I wrote above.

### LINKED ISSUE #1591: "INSTALL" button clipped or missing for some Boards/Library Manager items at minimum panel width
### Describe the problem

The Arduino IDE Library Manager is used to install/update/uninstall any of thousands of libraries. Each entry includes an "**INSTALL**" button used to install the version selected from the version menu.

This view is located in the left side panel, the width of which is adjustable.

🐛 When the panel is sized to the minimum width, the "**INSTALL**" button of certain entries is either clipped or completely hidden by the bottom border of the entry.

### To reproduce

1. Maximize the IDE window.
   **ⓘ** See the information under the "Additional context" section to learn about the relation of the problem to window dimensions.
1. Open the "**Library Manager**" view.
1. In the "**Filter your search...**" field, type `tensorflow`
1. Hover the mouse pointer over the right border of the left side panel.
1. Click and drag leftward to resize the panel to the minimum width.

🐛 The "**INSTALL**" button and even version menu of certain entries is either clipped or completely hidden by the bottom border of the entry.

image

image

### Expected behavior

The "**INSTALL**" button is always completely visible on the Library Manager entries.

### Arduino IDE version

2.0.1

### Operating system

- Windows
- Linux

### Operating system version

- Windows 10
- Ubuntu 20.04

### Additional context

I find that the problem is dependent on window dimensions/scaling. The problem occurs at the default 100% scale on my 1920x1200 display with the window maximized. On my 1366x768 display, the problem only occurs at scaling values of 60% and lower with the window maximized. With smaller window dimensions, I also must use smaller scaling values in order to produce the problem.

---

The same issue also occurs in a search for `irremote` or `sigfox`.

---

Related:

- https://github.com/arduino/arduino-ide/issues/1403
- https://github.com/arduino/arduino-ide/pull/1568

---

Additional reports:

- https://forum.arduino.cc/t/cant-install-boards-in-arduino-ide-2-0-3-no-install-button/1091088
- https://forum.arduino.cc/t/library-manager-panel-bug/1032762/4
- https://forum.arduino.cc/t/arduino-ide-2-unable-to-install-esp8266-board-definitions/1087325
- https://forum.arduino.cc/t/2-0-3-boards-manager-text-size-wonkiness/1064836
- https://forum.arduino.cc/t/cant-install-rp2040-boards/1059783/1
- https://forum.arduino.cc/t/cant-install-rp2040-boards/1059783/4
- https://github.com/arduino/arduino-ide/issues/1713
- https://github.com/arduino/arduino-ide/issues/1698
- https://forum.arduino.cc/t/esptool-exe-quarantined/1095560/4

---

#### Workaround

1. Open the Boards Manager or Library Manager view if it is not already open.
1. Hover your mouse pointer over the right border of the Boards/Library Manager view until the pointer changes to the horizontal resize double arrow.
1. Click and drag the mouse pointer to the right to make the Boards/Library Manager view wider.

After doing that, you should see the **INSTALL** button.

### Issue checklist

- [X] I searched for previous reports in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the problem still occurs when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My report contains all necessary details

### LINKED ISSUE #1607: Board menu should show 'update' not 'install' for updatable boards
### Describe the request

Currently the board menu will show an 'install' button for any board. If you have the current version of the board it will show the install button for the previous version, and if you are behind the current version it will also show an install button. The only way to differentiate the two is to cross check the two verison numbers.

Its much better to show an 'update' button for boards that can be updated, it makes the IDE easier to use because you can instantly see in the boards manager if a board is updatable instead of having to check version numbers. It also differentiates the action of updating something from that of installing an older version over the current version.

### Describe the current behavior

Boards that are updateable do not have a button that says 'update'

### Arduino IDE version

2.0.1

### Operating system

Windows

### Operating system version

10

### Additional context

_No response_

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My request contains all necessary details

### LINKED ISSUE #1707: Library Manager version menu selection changes if user closes dependencies installation dialog
### Describe the problem

An Arduino library can specify dependencies on other libraries via its [metadata](https://arduino.github.io/arduino-cli/dev/library-specification/#library-metadata). When the user installs a library with defined dependencies via Library Manager, Arduino IDE shows a dialog that offers to install these dependencies along with the library.

image

If the user closes that dialog, the selection in the version menu changes:

- If the latest version is currently selected, the previous version is selected
- If a previous version is currently selected, the latest version is selected

🐛 The user will likely not notice this unexpected change. If they trigger an install operation on the library once more, they will install a different version than they intended.

If the version the IDE attempts to switch the menu to is currently installed, the current selection is retained, but an install operation will attempt to install the version it attempted to switch the menu to.

🐛 The user will be confused when the installation operation fails due to attempting to install a different version than was already selected.

### To reproduce

1. Find the "**ArduinoModbus**" library in the **Library Manager** view.
1. If the library is already installed, uninstall it.
1. Take note of the selection in the version menu (e.g., "**1.0.7**")
1. Click the **INSTALL** button.
1. Click the **X** icon on the "**Install library dependencies**" dialog.
   🐛 The selection in the version menu changed.
1. Select the version of the library before the latest version from the menu (e.g., if the latest version is 1.0.7, select "**1.0.6**" from the menu).
1. Click the **INSTALL** button and install the library.
   🙂 The version menu selection changes to the latest version of the library.
1. Click the **INSTALL** button.
1. Click the **X** icon on the "**Install library dependencies**" dialog.
   🙂 The latest version of the library is still selected in the version menu.
1. Click the **INSTALL** button and install the library.
   🐛 The version of the library selected from the version menu is not installed. The Output view shows that the install operation attempted to install the previous version:
   > `Already installed ArduinoModbus@1.0.6`

### Expected behavior

Version menu selection only changes after a successful installation.

### Arduino IDE version

2.0.2

### Operating system

Windows

### Operating system version

10

### Issue checklist

- [X] I searched for previous reports in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the problem still occurs when using the latest [nightly build](https://www.arduino.cc/en/software#nightly-builds)
- [X] My report contains all necessary details

### LINKED ISSUE #1924: Board/Library Manager "Versions" not intutitive.
### Describe the request

The behavior of the version drop down immediately changing (dropping the installing version) and highlighting the prior version after clicking install does not seem very intuitive at first use, it tripped me up initially before recognizing what it was doing.

### Describe the current behavior

When you select to install the latest version of a library or board plugin, the version drop-down immediately changes to the prior version.  This can make the user think they are installing the older version.  You have no visual indication of "Installing version x.x" in the Manager window, the user would have to be monitoring the Output window.

Once it is installed of course, the "Version x.x Installed" appears, but is separated by the description from the version drop-down and install buttons.  The drop-down the no longer lists the installed version.

I feel it would be more intuitive if the latest version remained in the drop-down, with "Installing or Installed" on the same line when the drop-down is expanded, but of course then the install button should be grayed out if the installed version is selected.

Just a thought/experience I wanted to share..

### Arduino IDE version

2.0.4

### Operating system

N/A

### Operating system version

n/a

### Additional context

_No response_

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://www.arduino.cc/en/software#nightly-builds)
- [X] My request contains all necessary details

### LINKED ISSUE #1941: `focus` the boards/library mananger `<input>` when the widget opens
### Describe the problem

When I open the library or board manager widget, I want to have the `focus` on the `<input>` so that I can type without clicking anywhere. It's a regression.

2.0.0:

https://user-images.githubusercontent.com/1405703/223446392-70789c2c-6ee5-4354-98f4-3e5ff3e4878b.mp4

2.0.4:

https://user-images.githubusercontent.com/1405703/223446463-d487c772-b6b7-4f14-aa5d-e13b2f1f2bc2.mp4

### To reproduce

 - create a new sketch (or reset the workbench layout),
 - open the library or boards manager widget,
 - the `focus` is not on the `<input>` 

### Expected behavior

2.0.0 behavior

### Arduino IDE version

2.0.4

### Operating system

macOS

### Operating system version

12.6.3

### Additional context

_No response_

### Issue checklist

- [X] I searched for previous reports in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the problem still occurs when using the latest [nightly build](https://www.arduino.cc/en/software#nightly-builds)
- [X] My report contains all necessary details

### PR DESCRIPTION
**Pull request by kittaakos**
### Change description

TODOs:
 - [x] https://github.com/arduino/arduino-ide/issues/19
   - When the installed version is selected in the version dropdown, IDE2 shows the `Remove` button.
   - When a library is installed, hovering over the installed version, `REMOVE` is visible.
   - Removing a library from the `...` context menu is also possible.

https://user-images.githubusercontent.com/1405703/224006766-3737d8bc-3210-4541-b82e-bf4e56954553.mp4

https://user-images.githubusercontent.com/1405703/224006834-486308e1-a451-4d0b-95d1-af4fc1565afe.mp4

 - [x] https://github.com/arduino/arduino-ide/issues/781
   - Open the libraries manager widget,
   - Search for `Arduino Uno WiFi Dev Ed Library`,
   - If the library is installed, uninstall it,
   - Verify that only `0.0.3` is available in the select,
   - Click on the `Install` button,
   - Library is installed, but the select still shows `0.0.3` version, and the button shows `Remove` instead of `Install`,
   - Click on `...`, and the context menu shows `Remove`. `Remove` from the context menu is always visible when the library is installed. It does not depend on the button's state.

https://user-images.githubusercontent.com/1405703/224006904-242ba613-2790-434a-a260-0dc2f72dbe75.mp4

 - [x] https://github.com/arduino/arduino-ide/issues/1591
   - The version `<select>` and the install `<button>` were made to be always in #1568, so there was not much to fix, but remove the bogus cell height calculation and switch to the `react-virtouso` library from Theia.
 - [x] https://github.com/arduino/arduino-ide/issues/1607
   - When a library is installed, but the installed version is not the latest, and the latest version is selected in the dropdown, IDE2 shows `Update` instead of `Install`

https://user-images.githubusercontent.com/1405703/224005054-f464fe48-2935-4541-b836-49cd688748fb.mp4

 - [x] https://github.com/arduino/arduino-ide/issues/1679
   - Create a new sketch,
   - Open the libraries manager widget,
   - Search for `arduinoiotcloud`,
   - Select version `1.10.0`,
   - Click on the `Install` button,
   - Dialog pops,
   - You can navigate with <kbd>Shift</kbd>+<kbd>Tab</kbd> (on macOS).

https://user-images.githubusercontent.com/1405703/223973556-9d8a0030-2e4f-4772-bdcc-630fcfbcd619.mp4

 - [x] https://github.com/arduino/arduino-ide/issues/1707
   - Create a new sketch,
   - Open the libraries manager widget,
   - Search for `arduinoiotcloud`,
   - Select version `1.10.0`,
   - Click on the `Install` button,
   - Dialog pops, close the dialog with the `X`,
   - Verify that the version `select` still shows `1.10.1`.

https://user-images.githubusercontent.com/1405703/223971468-6b019f84-9790-4672-bede-960031de5b5a.mp4

 - [x] https://github.com/arduino/arduino-ide/issues/1924
   - IDE2 always selects the latest available version by default,
   - When toggling/showing the view, all version selects reset to the default:

https://user-images.githubusercontent.com/1405703/224002986-38f9dd74-e737-45c1-a849-790b8001fed5.mp4

   - When changing the version of a lib, all other versions reset to the default:

https://user-images.githubusercontent.com/1405703/224003386-35fa5a99-b126-438c-990f-d1e21a5ea109.mp4

   - When installing or uninstalling a library, the version selection of all libraries reset to the default.

https://user-images.githubusercontent.com/1405703/224004076-c7566a94-070d-4d0c-a1a6-5edcb8e62afb.mp4

https://user-images.githubusercontent.com/1405703/224004168-453e0ee1-15d5-43df-83ec-527135f79776.mp4

   - All lib versions reset to the default when changing the query term or the search filters.

 - [x] https://github.com/arduino/arduino-ide/issues/1941
   - Create a new sketch (or reset the workbench layout),
   - Open the libraries/boards manager,
   - The `focus` is on the `<input>` so that you can type to filter without clicking anywhere.

https://user-images.githubusercontent.com/1405703/223965716-598794e0-7a88-4e82-8fcb-50412f98dded.mp4

### Other information

 - The `Install`/`Update`/`Remove` button is disabled when installing, updating, or removing a library.

https://user-images.githubusercontent.com/1405703/224007676-47ea8e46-ea5b-4e28-8fe3-c48b44ab3e7d.mp4

 - Colors work well with all three supported themes.

https://user-images.githubusercontent.com/1405703/224008082-f67470d9-e3d9-4610-9fcd-c906acc148cd.mp4

 - When a library is installed with at least one example, the `Examples` submenu is available from the `...` context menu.

https://user-images.githubusercontent.com/1405703/224008508-7cd7a905-f12a-4608-bebe-d29b8fcdda3f.mp4

 - Can open `More info` from the context menu.
 - Can select another version from the context menu. The menu is not visible when there is only one version available.

Screen Shot 2023-03-09 at 12 21 53

### Reviewer checklist

Closes #19
Closes #781
Closes #1591
Closes #1607
Closes #1707
Closes #1924
Closes #1941

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: arduino-ide-extension/src/browser/widgets/component-list/list-item-renderer.tsx [hunk 1]
```diff
@@ -1,137 +1,779 @@
...
+import {
+  ArduinoMenus,
+  examplesLabel,
+  unregisterSubmenu,
+} from '../../menu/arduino-menus';
+
+const moreInfoLabel = nls.localize('arduino/component/moreInfo', 'More info');
+const otherVersionsLabel = nls.localize(
+  'arduino/component/otherVersions',
+  'Other versions'
```
Comment (per1234) [inline arduino-ide-extension/src/browser/widgets/component-list/list-item-renderer.tsx:49]:
```suggestion
  'Other Versions'
```
The IDE's other context menus use title case for the menu items so I assume that is the standardized style.

**Review by per1234 — CHANGES_REQUESTED** [review]

File: arduino-ide-extension/src/browser/widgets/component-list/list-item-renderer.tsx (hunk 1 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/widgets/component-list/list-item-renderer.tsx]: It's in lowercase in the design, but I agree to change it to uppercase.

https://github.com/arduino/arduino-ide/pull/1927/commits/8c973ab111d42dbbf4dca753d6f1658739d63de0

**Review by per1234 — APPROVED** [review]
Thanks Akos!

**Comment by 91volt** [comment]
I believe it would be beneficial to keep the "More info" link within the card, as it provides users with an easy and accessible way to access additional information: with the introduction of the description truncation for a more compact layout, having this link easily discoverable would be especially valuable. 
Proposed solution:
image


Additionally, although it wasn't in the proposed design, I think it would be a great to show the contextual menu icons when the user hovers over the card, similar to how we currently use it for sketches listed in the sketchbook sidebar panel. This would provide users with a cleaner card while them are not interacting with the content. What are your thoughts on this? 

https://user-images.githubusercontent.com/92443068/224663890-0ee061e3-c76b-4b1a-ba51-0392273e32db.mov

**Comment by kittaakos** [comment]
> 1. I believe it would be beneficial to keep the "More info" link within the card,

It's a bug. I do not know how I missed it. Thanks!

> 2\. I think it would be better to show the contextual menu icons when the user hovers over the card

Sure. I wanted to do it as you're requesting now, but it was different in the design.

**Comment by kittaakos** [comment]
> I have only 2 points (not critical imho)

Thanks for the feedback. Both issues should be fixed.
Missing `More info`:
Screen Shot 2023-03-13 at 11 19 50

Screen Shot 2023-03-13 at 10 57 14

Show `...` context menu on hover:

https://user-images.githubusercontent.com/1405703/224674484-74d50003-4206-4bac-80c6-f46dd225293b.mp4

**Review by 91volt — APPROVED** [review]

**Pull request merged**